libnetwork/driver_plugin.py

```diff
@@ -0,0 +1,280 @@
+
+from pycalico.util import generate_cali_interface_name
+from subprocess32 import CalledProcessError
+from werkzeug.exceptions import HTTPException, default_exceptions
+from netaddr import IPNetwork
+from pycalico.datastore_datatypes import IF_PREFIX, Endpoint, IPPool
+from pycalico import netns
+
+from datastore_libnetwork import LibnetworkDatastoreClient
+
+FIXED_MAC = "EE:EE:EE:EE:EE:EE"
+CONTAINER_NAME = "libnetwork"
+ORCHESTRATOR_ID = "libnetwork"
+
+# How long to wait (seconds) for IP commands to complete.
+IP_CMD_TIMEOUT = 5
+
+hostname = socket.gethostname()
+client = LibnetworkDatastoreClient()
+
+# Return all errors as JSON. From http://flask.pocoo.org/snippets/83/
+# This ensures that uncaught exceptions get returned to libnetwork in a useful
+# way
+def make_json_app(import_name, **kwargs):
+    """
+    Creates a JSON-oriented Flask app.
+
+    All error responses that you don't specifically
+    manage yourself will have application/json content
+    type, and will contain JSON like this (just an example):
+
+    { "Err": "405: Method Not Allowed" }
+    """
+    def make_json_error(ex):
+        response = jsonify({"Err": str(ex)})
+        response.status_code = (ex.code
+                                if isinstance(ex, HTTPException)
+                                else 500)
+        return response
+
+    wrapped_app = Flask(import_name, **kwargs)
+
+    for code in default_exceptions.iterkeys():
+        wrapped_app.error_handler_spec[None][code] = make_json_error
+
+    return wrapped_app
+
+app = make_json_app(__name__)
+app.logger.addHandler(logging.StreamHandler(sys.stdout))
+app.logger.setLevel(logging.DEBUG)
+app.logger.info("Application started")
+
+# The API calls below are documented at
+# https://github.com/docker/libnetwork/blob/master/docs/remote.md
+
+
+@app.route('/Plugin.Activate', methods=['POST'])
+def activate():
+    json_response = {"Implements": ["NetworkDriver"]}
+    app.logger.debug("Activate response JSON=%s", json_response)
+    return jsonify(json_response)
+
+
+@app.route('/NetworkDriver.GetCapabilities', methods=['POST'])
+def get_capabilities():
+    json_response = {"Scope": "global"}
+    app.logger.debug("GetCapabilities response JSON=%s", json_response)
+    return jsonify(json_response)
+
+
+@app.route('/NetworkDriver.CreateNetwork', methods=['POST'])
+def create_network():
+    # force is required since the request doesn't have the correct mimetype
+    # If the JSON is malformed, then a BadRequest exception is raised,
+    # which returns a HTTP 400 response.
+    json_data = request.get_json(force=True)
+    app.logger.debug("CreateNetwork JSON=%s", json_data)
+
+    # Create the CNM "network" as a Calico profile.
+    network_id = json_data["NetworkID"]
+    app.logger.info("Creating profile %s", network_id)
+    client.create_profile(network_id)
+
+    # Create a calico Pool for the CNM pool that was passed in.
+    for version in (4, 6):
+        if json_data["IPv%sData" % version]:
```

Comment:
minor niggle:  would be nicer to just get the ip data once, e.g.
ip_data = json_data[blah]
if ip_data then do stuff with ip_data